### PR TITLE
Add role for setup of kerberos

### PR DESCRIPTION
**Pull request by ezr-ondrej**
Role for setting up of kerberos with an existing domain AD controller.
Documentation for such setup in easiest local scenario.

Comment:
Mostly for my future reference I guess.

I was thinking about putting the documentation in `docs/advanced` and just linking it, but unsure what's better.

I can actually remove the playbook as it is quite simple, move the whole documentation to the role `README.md` and just link there from development.

### CODE REVIEW

File: docs/development.md [hunk 1]
```diff
@@ -348,9 +349,82 @@ In the vagrant box, the dynflow repository is cloned to `/home/vagrant/dynflow`.
...
+This list should be organized by foreman feature.
+
+* [Kerberos](#foreman-with-ad)
+
+### Foreman with AD
+
+#### Prepare AD controller
+
+You need the AD controller with domain controller you want to test this against.
+Easiest is to install Windows server (ideally core to safe resources, but its harder to install AD and setup DNS).
```
Comment (ekohl) [inline docs/development.md:370]:
```suggestion
Easiest is to install Windows server (ideally core to save resources, but its harder to install AD and setup DNS).
```

File: docs/development.md [hunk 2]
```diff
@@ -348,9 +349,82 @@ In the vagrant box, the dynflow repository is cloned to `/home/vagrant/dynflow`.
...
 5. From the smoker box, run tests as the vagrant user using the alias or running pytest commands manually. To change the testing options, please see [the smoker documentation](https://github.com/theforeman/smoker) and modify the alias or manually run pytest commands as necessary.
 
+## Advanced features
+
+These are playbooks, roles and variables provided to use additional funcionalities to test advanced foreman features.
+This list should be organized by foreman feature.
+
+* [Kerberos](#foreman-with-ad)
+
+### Foreman with AD
```
Comment (ekohl) [inline docs/development.md:365]:
```suggestion
### Foreman with Active Directory
```

File: docs/development.md [hunk 3]
```diff
@@ -348,9 +349,82 @@ In the vagrant box, the dynflow repository is cloned to `/home/vagrant/dynflow`.
...
+
+You need the AD controller with domain controller you want to test this against.
+Easiest is to install Windows server (ideally core to safe resources, but its harder to install AD and setup DNS).
+This controller doesn't need to be searchable in your environmnet, we will point libvirt there.
+
+I will be using domain ```example.com``` and expect your AD controller runs on IP ```192.168.1.8```.
+
+#### Setup libvirt
+
+Here we will edit the libvirt network, so its dnsmasq point the domain request to our controller.
```
Comment (ekohl) [inline docs/development.md:377]:
```suggestion
Here we will edit the libvirt network, so its dnsmasq points the domain request to our controller.
```

File: docs/development.md [hunk 4]
```diff
@@ -348,9 +349,82 @@ In the vagrant box, the dynflow repository is cloned to `/home/vagrant/dynflow`.
...
+    <dnsmasq:option value="server=/example.com/192.168.1.8"/>
+  </dnsmasq:options>
+</network>
+
+```
+
+#### Setup forklift machine
+
+Then add the machine and specify the domain, playbook and password for the join user.
+I'm using short name as longer names tend to have issues with NETBIOS names that end up shortened, and hostname != NETBIOS name could cause issues.
```
Comment (ekohl) [inline docs/development.md:396]:
```suggestion
I'm using short names as longer names tend to have issues with NETBIOS names that end up shortened, and hostname != NETBIOS name could cause issues.
```

File: docs/development.md [hunk 5]
```diff
@@ -348,9 +349,82 @@ In the vagrant box, the dynflow repository is cloned to `/home/vagrant/dynflow`.
...
+If you choose longer name you're on your own, contributions with fixes welcome! :)
+
+```yaml
+# boxes.d/99-local.yaml
+
+foreman-ad:
+  box: centos7
+  memory: 4098
```
Comment (ekohl) [inline docs/development.md:404]:
```suggestion
  memory: 4096
```

File: docs/development.md [hunk 6]
```diff
@@ -348,9 +349,82 @@ In the vagrant box, the dynflow repository is cloned to `/home/vagrant/dynflow`.
...
+#### Setup forklift machine
+
+Then add the machine and specify the domain, playbook and password for the join user.
+I'm using short name as longer names tend to have issues with NETBIOS names that end up shortened, and hostname != NETBIOS name could cause issues.
+If you choose longer name you're on your own, contributions with fixes welcome! :)
+
+```yaml
+# boxes.d/99-local.yaml
+
+foreman-ad:
```
Comment (ekohl) [inline docs/development.md:402]: Perhaps this should also be in the example boxes file?

File: roles/foreman_installer/tasks/module_local.yml [hunk 7]
```diff
@@ -0,0 +1,35 @@
+---
```
Comment (ekohl) [inline roles/foreman_installer/tasks/module_local.yml:1]: Why is this file needed?

File: roles/realm_join/tasks/realm_join_ad.yml [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+  loop:
+    - sssd
+    - adcli
+    - realmd
+    - ipa-python-compat
+    - krb5-workstation
+    - samba-common-tools
+    # Irrelevant, missing installer support for gssproxy
+    - gssproxy
+    - nfs-utils
```
Comment (ekohl) [inline roles/realm_join/tasks/realm_join_ad.yml:34]: I think `name` also accepts a list and that's faster than looping.

File: roles/realm_join/tasks/realm_join_ad.yml [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+  become: yes
+  package: name=pexpect state=latest
+  when: ansible_distribution != "CentOS" or ansible_distribution_major_version >= '8'
+
+- name: Install pip to get supported pexpect version
+  become: yes
+  package: "name={{ item }} state=present"
+  loop:
+    - epel-release
+    - python-pip
```
Comment (ekohl) [inline roles/realm_join/tasks/realm_join_ad.yml:12]: I think `name` accepts a list as well and is faster than looping.

File: roles/realm_join/tasks/realm_join_ad.yml [hunk 10]
```diff
@@ -0,0 +1,116 @@
+---
+- name: Ensure pexpect to answer password
+  become: yes
+  package: name=pexpect state=latest
```
Comment (ekohl) [inline roles/realm_join/tasks/realm_join_ad.yml:4]: Writing it out is generally cleaner (repeated below, not going to repeat the comment)
```suggestion
  package:
    name: pexpect
    state: latest
```

File: roles/realm_join/files/httpd-gssproxy.service [hunk 11]
```diff
@@ -0,0 +1,3 @@
+.include /lib/systemd/system/httpd.service
```
Comment (ekohl) [inline roles/realm_join/files/httpd-gssproxy.service:1]: `.include` isn't needed if you create it as a drop in. That means storing it as `/etc/systemd/system/httpd.service.d/gss-use-proxy.conf` (or similar). On newer systemd versions (included in EL8) the `.include` statement generates a deprecation warning.

**Review by ekohl — CHANGES_REQUESTED** [review]
Thanks! Looks like a nice thing to have.

File: roles/realm_join/files/httpd-gssproxy.service (hunk 11 above)
Comment (ezr-ondrej) [reply roles/realm_join/files/httpd-gssproxy.service:1]: I'm trying to follow the Satellite manual as strictly as possible - that is why I'm having the `gssproxy` steps even though those are not used.
Noted for the manual improvements tho! Thanks!

File: roles/foreman_installer/tasks/module_local.yml (hunk 7 above)
Comment (ezr-ondrej) [reply roles/foreman_installer/tasks/module_local.yml:1]: :shy:

File: roles/realm_join/files/httpd-gssproxy.service (hunk 11 above)
Comment (ekohl) [reply roles/realm_join/files/httpd-gssproxy.service:1]: We should file a docs bug then.

File: docs/development.md (hunk 6 above)
Comment (ezr-ondrej) [reply docs/development.md:402]: I don't think so, this is very obscure setup for most developers (you need the AD controller first), so it is hardly something most users would use, so I think it should not be in examples of everyuser (and thus being offered in `vagrant up` autocomplete if user doesn't delete it)

File: docs/development.md (hunk 4 above)
Comment (ezr-ondrej) [reply docs/development.md:396]: I believe my version is correct.

File: docs/development.md (hunk 6 above)
Comment (ekohl) [reply docs/development.md:402]: https://github.com/theforeman/forklift/blob/master/vagrant/boxes.d/99-local.yaml.example isn't loaded by default so it wouldn't show up in the autocomplete after a fresh clone. However, I'm fine either way since it is indeed not something that works on its own.

File: docs/development.md (hunk 4 above)
Comment (ekohl) [reply docs/development.md:396]: I think it's either `a short name` or `short names`, but since we're both not native speakers we may want to ask another reviewer :)

File: docs/development.md (hunk 6 above)
Comment (ezr-ondrej) [reply docs/development.md:402]: I know, but first thing most people will do is probably `cp 99-local.yaml.example 99-local.yaml` to try those examples out :shrug:
At least that's what I've done the first time.

File: roles/realm_join/tasks/realm_join_ad.yml (hunk 9 above)
Comment (ezr-ondrej) [reply roles/realm_join/tasks/realm_join_ad.yml]: :+1: yes it indeed does!

File: docs/development.md (hunk 4 above)
Comment (ezr-ondrej) [reply docs/development.md]: Right! I'm missing the `a`. I was just refering to singular being used correctly there :)

File: roles/realm_join/files/httpd-gssproxy.service (hunk 11 above)
Comment (ezr-ondrej) [reply roles/realm_join/files/httpd-gssproxy.service:1]: I'll followup with the improvements of the feature, docs improvement is part of it, I will take this into account :+1:

File: roles/realm_join/tasks/realm_join_ad.yml [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+- name: Create httpd conf dir to put keytab [Hack - remove once using gssproxy]
+  become: yes
+  file:
+    path: /etc/httpd/conf
+    state: directory
+    mode: '0775'
+    recurse: yes
+
+- name: Check wheter the keytab exists
```
Comment (ekohl) [inline roles/realm_join/tasks/realm_join_ad.yml:68]:
```suggestion
- name: Check whether the keytab exists
```

File: roles/realm_join/tasks/realm_join_ad.yml [hunk 13]
```diff
@@ -0,0 +1,120 @@
+---
+- name: Ensure pexpect to answer password
+  become: yes
+  package:
+    name: pexpect
+    state: latest
+  when: ansible_distribution != "CentOS" or ansible_distribution_major_version >= '8'
```
Comment (ekohl) [inline roles/realm_join/tasks/realm_join_ad.yml:7]: So on RHEL 7 pexpect is present?

File: docs/development.md [hunk 14]
```diff
@@ -348,9 +349,82 @@ In the vagrant box, the dynflow repository is cloned to `/home/vagrant/dynflow`.
...
+This list should be organized by foreman feature.
+
+* [Kerberos](#foreman-with-active-directory)
+
+### Foreman with Active Directory
+
+#### Prepare AD controller
+
+You need the AD controller with domain controller you want to test this against.
+Easiest is to install Windows server (ideally core to save resources, but its harder to install AD and setup DNS).
```
[the remark this answers is not in the document]
Comment (ezr-ondrej) [reply docs/development.md:370]: Well, that would be indead cool, but we could just have freeIPA kerberos :)

File: roles/realm_join/tasks/realm_join_ad.yml (hunk 13 above)
Comment (ezr-ondrej) [reply roles/realm_join/tasks/realm_join_ad.yml:7]: I'd expect no ;) ... I'll add RHEL :)

File: docs/development.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (ezr-ondrej) [reply docs/development.md:370]: But let's first get it in without it... if there I find out people struggle to get some reviews done, I'll try to come up with the REALM installation here as well.

File: roles/realm_join/files/httpd-gssproxy.service (hunk 11 above)
Comment (ekohl) [reply roles/realm_join/files/httpd-gssproxy.service:1]: I still don't like implementing this because it's not the way it's supposed to be used. By placing the file in `/etc/systemd/system/httpd.service` it means you need to have a migration in the future. Probably not a huge issue for most forklift deployments, but I did want to mention it.

**Comment by ekohl** [comment]
@evgeni could you take a look at this and see if I've missed anything?

File: docs/development.md [hunk 15]
```diff
@@ -348,9 +349,82 @@ In the vagrant box, the dynflow repository is cloned to `/home/vagrant/dynflow`.
...
+This controller doesn't need to be searchable in your environmnet, we will point libvirt there.
+
+I will be using the domain ```example.com``` and expect your AD controller runs on IP ```192.168.1.8```.
```
Comment (evgeni) [inline docs/development.md:373]:
```suggestion
I will be using the domain ```example.com``` and expect your AD controller runs on IP ```192.0.2.8```.
```

[rfc5735](https://tools.ietf.org/html/rfc5735) ;)

Comment (evgeni) [reply docs/development.md:373]: And I think you want singe backticks here, not three?

File: docs/development.md [hunk 16]
```diff
@@ -348,9 +349,82 @@ In the vagrant box, the dynflow repository is cloned to `/home/vagrant/dynflow`.
...
+Provision it
+
+```vagrant up foreman-ad```
+
+Test it
+
+```vagrant ssh foreman-ad```
+```kinit test-domain-user``` # initialize Kerberos with an existing user in the AD domain
+```klist``` # verify the above went smooth
+```curl -k -u : --negotiate https://foreman-ad.example.com/users/extlogin``` # substitute for your fqdn
```
Comment (evgeni) [inline docs/development.md:425]:
```suggestion
```console
$ vagrant ssh foreman-ad
$ kinit test-domain-user # initialize Kerberos with an existing user in the AD domain
$ klist # verify the above went smooth
$ curl -k -u : --negotiate https://foreman-ad.example.com/users/extlogin # substitute for your fqdn
```

File: roles/realm_join/tasks/realm_join_ad.yml [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+  package:
+    name:
+      - epel-release
+      - python-pip
+    state: present
+  when: ansible_os_family == "RedHat" and ansible_distribution_major_version < '8'
+
+- name: Ensure pexpect to answer password from pip
+  become: yes
+  pip: name=pexpect
```
Comment (evgeni) [inline roles/realm_join/tasks/realm_join_ad.yml:21]:
```suggestion
  pip:
    name: pexpect
```

File: roles/realm_join/tasks/realm_join_ad.yml [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+  when: ansible_os_family == "RedHat" and ansible_distribution_major_version < '8'
```
Comment (evgeni) [inline roles/realm_join/tasks/realm_join_ad.yml:22]:
```suggestion
  when:
    - ansible_os_family == "RedHat"
    - ansible_distribution_major_version is version('8', '<')
```

File: roles/realm_join/tasks/realm_join_ad.yml [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+  when: ansible_os_family != "RedHat" or ansible_distribution_major_version >= '8'
+
+- name: Install pip to get supported pexpect version
+  become: yes
+  package:
+    name:
+      - epel-release
+      - python-pip
+    state: present
+  when: ansible_os_family == "RedHat" and ansible_distribution_major_version < '8'
```
Comment (evgeni) [inline roles/realm_join/tasks/realm_join_ad.yml:17]:
```suggestion
  when:
    - ansible_os_family == "RedHat"
    - ansible_distribution_major_version is version('8', '<')
```

File: roles/realm_join/tasks/realm_join_ad.yml [hunk 20]
```diff
@@ -0,0 +1,121 @@
+---
+- name: Ensure pexpect to answer password
+  become: yes
+  package:
+    name: pexpect
+    state: latest
+  # on Fedora this will be True due to the major version. Not nice, but works.
+  when: ansible_os_family != "RedHat" or ansible_distribution_major_version >= '8'
```
Comment (evgeni) [inline roles/realm_join/tasks/realm_join_ad.yml:8]:
```suggestion
  when: ansible_os_family != "RedHat" or ansible_distribution_major_version is version('8', '>')
```

But there is no `pexpect` on my Fedora 33, only `python3-pexpect`. And neither is there such a package on Debian ;-)

File: roles/realm_join/tasks/realm_join_ad.yml [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+
+- name: Get the keytab
+  become: yes
+  environment:
+    KRB5_KTNAME: 'FILE:/etc/httpd/conf/http.keytab'
+  expect:
+    command: "net ads keytab add HTTP -U {{ foreman_realm_directory_admin_name }} -d3 -s /etc/net-keytab.conf"
+    responses:
+      (?i)password: '{{ foreman_realm_directory_admin_password }}'
+  when: not keytab_result.stat.exists
```
Comment (evgeni) [inline roles/realm_join/tasks/realm_join_ad.yml:82]: instead of `when`, you probably can have something like `creates: /etc/httpd/conf/http.keytab` (and then the step before can be remoted)

File: roles/realm_join/tasks/realm_join_ad.yml [hunk 22]
```diff
@@ -0,0 +1,121 @@
+---
+- name: Ensure pexpect to answer password
+  become: yes
+  package:
+    name: pexpect
+    state: latest
```
Comment (evgeni) [inline roles/realm_join/tasks/realm_join_ad.yml:6]: Does it really need to be `latest` (aka: trigger upgrade on re-run?)

**Review by evgeni — CHANGES_REQUESTED** [review]

File: docs/development.md (hunk 15 above)
Comment (ezr-ondrej) [reply docs/development.md:373]: Yeah, I guess I can, no idea why I used three, it was bit late and probably I went with three is more then one :thinking: :shrug:

File: roles/realm_join/tasks/realm_join_ad.yml (hunk 22 above)
Comment (ezr-ondrej) [reply roles/realm_join/tasks/realm_join_ad.yml:6]: Probably not, but quite a new version is necessary, so I didn't want to go with just simple `present` in case there is old version, but no idea if that can even happen on 8 :thinking:

File: roles/realm_join/tasks/realm_join_ad.yml [hunk 23]
```diff
@@ -0,0 +1,121 @@
+---
+- name: Ensure pexpect to answer password
+  become: yes
+  package:
+    name: pexpect
```
Comment (evgeni) [inline roles/realm_join/tasks/realm_join_ad.yml:5]: On EL8:
```console
[root@a8f349df7349 /]# yum install pexpect
Failed to set locale, defaulting to C.UTF-8
Last metadata expiration check: 0:00:08 ago on Tue Jan 12 12:57:58 2021.
No match for argument: pexpect
Error: Unable to find a match: pexpect
```

similar on Fedora 33 and Debian 10, where is this package coming from?

Should this be `python3-pexpect`?

File: roles/realm_join/tasks/realm_join_ad.yml [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+    state: latest
+  # on Fedora this will be True due to the major version. Not nice, but works.
+  when: ansible_os_family != "RedHat" or ansible_distribution_major_version is version('8', '>')
```
Comment (evgeni) [inline roles/realm_join/tasks/realm_join_ad.yml:8]: this evaluates to: "on debian" or "on all fedoras" or "on EL8+"
is that what you wanted?

**Review by evgeni — CHANGES_REQUESTED** [review]

File: roles/realm_join/tasks/realm_join_ad.yml (hunk 24 above)
Comment (ezr-ondrej) [reply roles/realm_join/tasks/realm_join_ad.yml:8]: probably not it I see now that it differs on debian and RHEL.

File: roles/realm_join/tasks/realm_join_ad.yml (hunk 23 above)
Comment (ezr-ondrej) [reply roles/realm_join/tasks/realm_join_ad.yml]: It should .... python3-pexpect on RedHat > 8 and Fedora and python-pexpect on Debian.

**Comment by ezr-ondrej** [comment]
I've extracted the pexpect install to separate file as it was cluttering the main file quite a bit already.

File: roles/realm_join/tasks/realm_join_ad.yml [hunk 25]
```diff
@@ -0,0 +1,96 @@
+---
+- include_tasks: install_pexpect.yml
```
Comment (evgeni) [inline roles/realm_join/tasks/realm_join_ad.yml:2]: this file is not in the PR? :)

File: roles/realm_join/tasks/main.yml [hunk 26]
```diff
@@ -0,0 +1 @@
+- include_tasks: realm_join_ad.yml
```
Comment (evgeni) [inline roles/realm_join/tasks/main.yml:1]: what's the reason to have all tasks in another file? prep-work for `realm_join_ipa.yml`?

File: roles/realm_join/tasks/realm_join_ad.yml (hunk 25 above)
Comment (ezr-ondrej) [reply roles/realm_join/tasks/realm_join_ad.yml:2]: :facepalm:

File: roles/realm_join/tasks/install_pexpect.yml [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+  when:
+    - ansible_os_family == "RedHat"
+    - ansible_distribution_major_version is version('8', '>')
+
+- name: Ensure pexpect to answer password
+  become: yes
+  package:
+    name: python-pexpect
+    state: latest
+  # on Fedora this will be True due to the major version. Not nice, but works.
```
Comment (evgeni) [inline roles/realm_join/tasks/install_pexpect.yml:16]: I hightly doubt that `ansible_os_family: Debian` will ever be true on Fedora ;)

I think this comment can co now.

Comment (ezr-ondrej) [reply roles/realm_join/tasks/install_pexpect.yml]: xD yes, indeed ....

File: roles/realm_join/tasks/main.yml (hunk 26 above)
Comment (ezr-ondrej) [reply roles/realm_join/tasks/main.yml:2]: yeah, doesn't make much sense xD I think first I was thinking it could run the installer afterwards, but then decided against it.
So now it's tree steps, get the pexpect on the necessary version, join the realm and then prepare the keytabs for apache.

File: roles/realm_join/tasks/install_pexpect.yml [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+      - epel-release
+      - python-pip
+    state: present
+  when:
+    - ansible_os_family == "RedHat"
+    - ansible_distribution_major_version is version('8', '<')
+
+- name: Ensure pexpect to answer password from pip
+  become: yes
+  pip:
```
Comment (evgeni) [inline roles/realm_join/tasks/install_pexpect.yml:32]: pexpect 2.3-11.el7 that is in EL7 is too old I guess?

File: roles/realm_join/tasks/install_pexpect.yml [hunk 29]
```diff
@@ -0,0 +1,36 @@
+- name: Ensure pexpect to answer password
+  become: yes
+  package:
+    name: python3-pexpect
```
Comment (evgeni) [inline roles/realm_join/tasks/install_pexpect.yml:4]: instead of having two blocks for different OSes, we could have vars like this:

https://github.com/theforeman/forklift/blob/4090e636e55d83a39d7f7cdaf5ce9b1f3ed41dc8/roles/sos_report/tasks/main.yml#L2-L8

https://github.com/theforeman/forklift/blob/4090e636e55d83a39d7f7cdaf5ce9b1f3ed41dc8/roles/sos_report/vars/Debian.yml#L2
https://github.com/theforeman/forklift/blob/4090e636e55d83a39d7f7cdaf5ce9b1f3ed41dc8/roles/sos_report/vars/RedHat.yml#L2

It's up to you if you want to go that route, for a single package it might be overkill.

File: roles/realm_join/tasks/install_pexpect.yml (hunk 28 above)
Comment (ezr-ondrej) [reply roles/realm_join/tasks/install_pexpect.yml:32]: Yes, we need `>= 3.3` for the ansible module to work.

File: roles/realm_join/tasks/install_pexpect.yml (hunk 29 above)
Comment (ezr-ondrej) [reply roles/realm_join/tasks/install_pexpect.yml:4]: I :heart: it. Lets do it

**Review by evgeni — APPROVED** [review]
ACK from an "the ansible role reads fine" POV, I didn't test it actually works or matches the docs.

**Review by ekohl — APPROVED** [review]
Thanks for addressing all concerns!
